contracts/BridgeExecutor.sol

```diff
@@ -0,0 +1,139 @@
+//SPDX-License-Identifier: UNLICENSED
+
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "./lib/Pausable.sol";
+import "./SharedStructs.sol";
+import "./lib/BoolTokenTransfer.sol";
+import "./access/AdminRole.sol";
+import "./access/BridgeRole.sol";
+
+contract BridgeExecutor is Initializable, Pausable, BridgeRole {
+    using BoolTokenTransfer for IERC20;
+
+    /*========================= CONTRACT STATE =========================*/
+    uint128 public constant MIN_GAS_LIMIT_FOR_SC_CALL = 10_000_000;
+    uint128 public constant DEFAULT_GAS_LIMIT_FOR_REFUND_CALLBACK = 20_000_000;
+    uint256 private lowestTxId;
+    uint256 private currentTxId;
+
+    mapping(uint256 => MvxTransaction) private pendingTransactions;
+
+    /*========================= PUBLIC API =========================*/
+    function initialize() public initializer {
+        __BridgeRole_init();
+        __Pausable_init();
+    }
+
+    function __BridgeProxy__init_unchained() internal onlyInitializing {
+        lowestTxId = 0;
+        currentTxId = 0;
+    }
+
+    function deposit(MvxTransaction calldata txn) external payable whenNotPaused onlyBridge returns (bool) {
+        pendingTransactions[currentTxId] = txn;
+        currentTxId++;
+
+        IERC20 token = IERC20(txn.token);
+        bool approvalSuccess = token.approve(txn.recipient, txn.amount);
+
+        return approvalSuccess;
+    }
+
+    function execute(uint256 txId) external whenNotPaused {
+        require(txId < currentTxId, "BridgeExecutor: Invalid transaction ID");
+        MvxTransaction memory txn = pendingTransactions[txId];
+
```

Comment:
We should also have a call that txn exists (it was not deleted). In case of reentrancy, when reaching back into this code, it will go on the else case, `_refundAndDeleteTxn(txId);`. Indeed it will try to transfer 0 back and the call will fail, but let's actually catch this behaviour and revert